Add support for Jar to Jar transformation to JarInfer

navahgar opened the pull request:
This adds support for annotating jar files, as opposed to creating an astubx file on the side. 

Summary of changes:

- Added support to annotate jar files with the annotations inferred by the analysis in JarInfer.
- Added a flag in JarInfer CLI to enable annotating jar files.
- Tested annotated jar files by checking for expected annotations.
- Added the expected annotations in the sources used for tests.

lazaroclapp commented on jar-infer/jar-infer-lib/src/test/java/com/uber/nullaway/jarinfer/AnnotationComparator.java:
```diff
@@ -0,0 +1,141 @@
+        classFile1.endsWith(".class") && classFile2.endsWith(".class"),
+        "invalid class files: " + classFile1 + " " + classFile2);
+    System.out.println("File1: " + classFile1);
+    System.out.println("File2: " + classFile2);
```
So, I know JarInfer is already pretty verbose with print debugging, so I am fine with keeping these for now, but I am not a fan of being that verbose with unconditional prints on landed code (even the `DEBUG` flag conditional prints are probably not the best way of handling debugging for this, but not 100% sure what to do about it)

navahgar replied:
This class is removed now.

lazaroclapp commented on jar-infer/jar-infer-lib/src/test/java/com/uber/nullaway/jarinfer/AnnotationComparator.java:
```diff
@@ -0,0 +1,141 @@
+import jdk.internal.org.objectweb.asm.tree.MethodNode;
+
+// This class is used to compare the method annotations on the given class / jar files.
+public class AnnotationComparator {
```
Is this class still used? Planned to be used? Or is it obsolete in favor of `AnnotationChecker`?

navahgar replied:
Removed the class.

lazaroclapp commented on jar-infer/jar-infer-lib/src/test/java/com/uber/nullaway/jarinfer/JarInferTest.java:
```diff
@@ -101,6 +103,54 @@ private void testJARTemplate(
+    DefinitelyDerefedParamsDriver.reset();
+    // TODO(ragr@): Package name has to be "" for the analysis to happen because in other cases, the
+    // class name
+    // does not have package as the prefix and hence it is ignored. Figure this out.
```
Not sure I understand this one...

navahgar replied:
This test is removed.

lazaroclapp commented on jar-infer/jar-infer-lib/src/test/java/com/uber/nullaway/jarinfer/JarInferTest.java:
```diff
@@ -255,11 +305,77 @@ public void toyNullTestAPI() throws Exception {
   }
 
+  private String getToyTestSrcWithExpectAnnotations() {
+    return "package toys;"
```
This code is weird to me. Should probably be multiple `addInputSource`: one for each interface and one for the class. Java won't let you define multiple top-level types in the same file, as far as I know. Although it seems it does work in the code below, um...

navahgar replied:
Removed the test.

lazaroclapp commented on jar-infer/test-java-lib-jarinfer/src/main/java/com/uber/nullaway/jarinfer/toys/Toys.java:
```diff
@@ -1,4 +1,13 @@
+import java.lang.annotation.RetentionPolicy;
+
+@Retention(RetentionPolicy.RUNTIME)
+@interface ExpectNullable {}
```
These should be in separate files

navahgar replied:
Keeping this as is, per our offline discussion.

lazaroclapp commented on jar-infer/test-java-lib-jarinfer/src/main/java/com/uber/nullaway/jarinfer/toys/Toys.java:
```diff
@@ -1,4 +1,13 @@
+@interface ExpectNullable {}
+
+@Retention(RetentionPolicy.RUNTIME)
+@interface ExpectNonnull {}
```
This one too.

navahgar replied:
Keeping this as is, per our offline discussion.

navahgar commented on jar-infer/jar-infer-lib/src/test/java/com/uber/nullaway/jarinfer/JarInferTest.java:
```diff
@@ -222,7 +272,7 @@ public void toyNonStatic() throws Exception {
   @Test
   public void toyJAR() throws Exception {
     testJARTemplate(
-        "com.uber.nullaway.jarinfer.toys.unannotated",
```
Added that back after our offline discussion. Thanks for clarifying.

navahgar commented on jar-infer/jar-infer-lib/src/main/java/com/uber/nullaway/jarinfer/BytecodeAnnotator.java:
```diff
@@ -0,0 +1,138 @@
+  }
+
+  public static final String javaxNullableDesc = "Ljavax/annotation/Nullable;";
+  public static final String javaxNonnullDesc = "Ljavax/annotation/Nonnull;";
```
I agree that is needed. Will do later.

navahgar commented on jar-infer/jar-infer-lib/src/main/java/com/uber/nullaway/jarinfer/BytecodeAnnotator.java:
```diff
@@ -0,0 +1,138 @@
+      Set<String> nullableReturns,
+      boolean DEBUG)
+      throws IOException {
+    BytecodeAnnotator.DEBUG = DEBUG;
```
Done.

navahgar commented on jar-infer/jar-infer-lib/src/main/java/com/uber/nullaway/jarinfer/BytecodeAnnotator.java:
```diff
@@ -0,0 +1,138 @@
+import org.objectweb.asm.Opcodes;
+
+// This class uses ASM to add the given annotations to methods in the given class / jar.
+public final class BytecodeAnnotator extends ClassVisitor implements Opcodes {
```
Done.

navahgar commented on jar-infer/jar-infer-lib/src/main/java/com/uber/nullaway/jarinfer/DefinitelyDerefedParamsDriver.java:
```diff
@@ -77,20 +80,25 @@ private static void LOG(boolean cond, String tag, String msg) {
 
   public static String lastOutPath = "";
   public static long analyzedBytes = 0;
+  public static long analysisStartTime = 0;
```
Leaving this as is, per our discussion offline.

navahgar commented on jar-infer/jar-infer-lib/src/main/java/com/uber/nullaway/jarinfer/DefinitelyDerefedParamsDriver.java:
```diff
@@ -140,127 +154,141 @@ public static Result run(String inPaths, String pkgName)
-        continue;
+      analyzeFile(pkgName, inPath);
+      if (DefinitelyDerefedParamsDriver.annotateBytecode) {
+        WriteAnnotations(inPath, outPath);
```
Clarified offline. Added a comment on the outPath parameter.

navahgar commented on jar-infer/jar-infer-lib/src/main/java/com/uber/nullaway/jarinfer/DefinitelyDerefedParamsDriver.java:
```diff
@@ -140,127 +154,141 @@ public static Result run(String inPaths, String pkgName)
-      IClassHierarchy cha = ClassHierarchyFactory.makeWithPhantom(scope);
-      Warnings.clear();
+    }
+    if (!DefinitelyDerefedParamsDriver.annotateBytecode) {
```
Clarified offline.

navahgar commented on jar-infer/jar-infer-lib/src/test/java/com/uber/nullaway/jarinfer/AnnotationChecker.java:
```diff
@@ -0,0 +1,99 @@
+
+// This class includes methods to check if the methods in the given class / jar files have
+// the expected annotations.
+public class AnnotationChecker {
```
Added.

navahgar commented on jar-infer/jar-infer-lib/src/test/java/com/uber/nullaway/jarinfer/AnnotationChecker.java:
```diff
@@ -0,0 +1,99 @@
+// This class includes methods to check if the methods in the given class / jar files have
+// the expected annotations.
+public class AnnotationChecker {
+  public static boolean CheckMethodAnnotationsInJar(
```
Done.

navahgar commented:
Thanks for the detailed comments. Addressed them. PTAL.

navahgar commented:
Summary of changes:
  * Added support to annotate jar files with the annotations inferred by the analysis in JarInfer.
  * Added a flag in JarInfer CLI to enable annotating jar files.
  * Tested annotated jar files by checking for expected annotations.
  * Added the expected annotations in the sources used for tests.

navahgar commented:
> Can you please add a more descriptive title and a quick summary of the top changes maybe to now we changed jarinfer

Updated the title and added a summary.

lazaroclapp commented on jar-infer/jar-infer-lib/src/test/java/com/uber/nullaway/jarinfer/AnnotationChecker.java:
```diff
@@ -14,47 +28,45 @@
+   * expected are specified in the form of a map. For example: map = {"ExpectNullable;",
+   * "Ljavax/annotation/Nullable;"} will check if all methods and params contain
+   * "Ljavax/annotation/Nullable;" iff "ExpectNullable;" is present.
+   */
```
`@param` and `@return` descriptions?

msridhar commented on jar-infer/jar-infer-lib/src/test/java/com/uber/nullaway/jarinfer/JarInferTest.java:
```diff
@@ -101,6 +103,24 @@ private void testJARTemplate(
+      Map<String, String> expectedToActualAnnotationsMap)
+      throws Exception {
+    String outputFolderPath = outputFolder.newFolder(pkg).getAbsolutePath();
+    DefinitelyDerefedParamsDriver.reset();
```
Might be better to rewrite `DefinitelyDerefedParamsDriver` to use instance methods rather than static, to get rid of this `reset()` method

lazaroclapp replied:
100% agreed. @navahgar, follow up PR for this?

navahgar replied:
Sure. That is the plan. I didn't want to make more changes to this already big diff.
